### PR TITLE
Add ManagedOSVersionChannel and skeleton for sync service

### LINKED ISSUE #63: CRD: `ManagedOSVersionChannel`
Part of https://github.com/rancher-sandbox/os2/issues/58

Create the following new CRDs and controller for the resources types described below
## `ManagedOSVersionChannel` and the controller

the `ManagedOSVersion` resources are populated by a controller, or either a kubernetes mechanism that satisfy `ManagedOSVersionChannel` source and constraints. The ManagedOSVersionChannel can be defined as such:

```
apiVersion: rancheros.cattle.io/v1
kind: ManagedOSVersionChannel
spec:
  type: "JSON"
  options: # map[string]interface{}
       url: "<file>.json"
```

where a type of channel (e.g. JSON, URL, XML, Github) populates `ManagedOSVersions` by using the specified options

It defines with which rules `ManagedOSVersion` should be populated (e.g. github repository tags, releases, a custom json file to parse)

The controller make sure to add labels/annotations where appropriately to notify that a ManagedOS version can be applied.

Ping the harvester team about their strategy and if it does cover their use cases too. Note they might have already a json file or something similar for their available versions, in which we might want to sync up within this card as we want to collect as much feedback as possible at this stage.

cc @bk201

### PR DESCRIPTION
**Pull request by mudler**
Fixes: https://github.com/rancher-sandbox/os2/issues/63

Signed-off-by: David Cassany <dcassany@suse.com>

### CODE REVIEW

File: pkg/operator/operator.go [hunk 1]
```diff
@@ -73,6 +83,11 @@ func Run(ctx context.Context, namespace string) error {
 	inventory.Register(ctx, clients)
 	registration.Register(ctx, clients)
 
-	aggregation.Watch(ctx, clients.Core.Secret(), namespace, "rancheros-operator", server.New(clients))
+	aggregation.Watch(ctx, clients.Core.Secret(), o.namespace, "rancheros-operator", server.New(clients))
+
+	for _, s := range o.services {
+		go s(ctx, clients)
```
Comment (mudler) [inline pkg/operator/operator.go]:
```suggestion
		go s(ctx, clients) //nolint:golint,errcheck
```

File: main.go [hunk 2]
```diff
@@ -37,7 +41,21 @@ func main() {
...
+	var ns []string
+	if os.Getenv("WATCH_NAMESPACE") != "" {
+		*namespaces = os.Getenv("WATCH_NAMESPACE")
+	}
+
+	if *namespaces != "" {
+		ns = strings.Split(*namespaces, ",")
+	}
+
+	//TODO check the proper namespace configuration, should it be the same namespace as the rancheros-operator?
```
Comment (davidcassany) [inline main.go:54]: I guess this TODO is obsolete, I think the current state is flexible and clear enough about namespaces

**Review by davidcassany — APPROVED** [review]
Looks good to me

**Comment by mudler** [comment]
There are three aspects here that I'd prefer to have as a separate follow up as they can all be carried in parallel:

- error events streamed back at resources. Currently errors are printed out in the operator logs, but ideally those should be streamed back as kubernetes events
- owner reference and cleanup orphan resources. At the moment `ManagedOSVersion` created by the controller don't have an owner set, so cleaning up a `ManagedOSVersionChannel` doesn't cleanup available versions
- hardcoded ticker interval. The ticker is now currently hardcoded, but this should be at least bumped to a higher value, and offer a way to optionally override this

**Comment by mudler** [comment]
Merging, created all necessary follow-ups

**Pull request merged**